File: exercicios/aula13/ex10_pipeline/parser.py
"""Exercício 10 — mini pipeline scanner -> tokens -> parser -> atributos -> AST.

Subconjunto MINIC implementado:

    programa  -> item*
    item      -> decl | cmd
    decl      -> tipo decl_item { , decl_item } ;      tipo -> int | float | bool
    decl_item -> id [ = expr ]
    cmd       -> bloco | if | while | return | atrib
    bloco     -> { item* }
    if        -> if ( expr ) cmd [ else cmd ]
    while     -> while ( expr ) cmd
    return    -> return [ expr ] ;
    atrib     -> id = expr ;
    expr      -> níveis || , && , == != , < <= > >= , + - , * / % , unário - ! ,
                 primária (id, int, float, true, false, ( expr ))

Atributos:

- **herdados**: `tipo` desce da palavra de tipo para cada decl_item
  (`decl_item.inh = tipo.type`, como no exercício 04); `escopo` desce do
  bloco para tudo o que está dentro dele (como no exercício 05). Os dois são
  parâmetros das funções do parser;
- **sintetizados**: `node` (a AST, construída pelas ações) e `tipo` de cada
  expressão, calculado a partir dos filhos com as regras do exercício 08
  (promoção int -> float; `erro` não gera diagnóstico em cascata).

Dependências relevantes: o tipo de um Id depende da declaração visível no
escopo herdado; o tipo de um Binary depende dos tipos dos dois filhos; a
checagem de uma atribuição depende do tipo do alvo e do valor. Tudo isso fica
disponível numa única passada da esquerda para a direita, então as ações
rodam durante a própria análise sintática.

Regras semânticas verificadas: identificador não declarado, redeclaração no
mesmo escopo (sombrear um escopo externo é permitido), compatibilidade de
tipos em inicialização e atribuição (mesmo tipo, ou int -> float) e condição
de if/while do tipo bool.

Limitações: sem funções, vetores, for, print/read nem char/string; a análise
para no primeiro erro de sintaxe (recuperação é o assunto do exercício 09).

Uso:  python parser.py [arquivo | -]      (sem argumento: lê a entrada padrão)
"""

import sys
from dataclasses import dataclass
from pathlib import Path

import ast_nodes as A
from scanner import TokenStream, scan

OK, ERRO_USO, ERRO_LEXICO, ERRO_SINTAXE, ERRO_SEMANTICO = 0, 1, 2, 3, 4
TIPOS = {"KW_INT": "int", "KW_FLOAT": "float", "KW_BOOL": "bool"}
NUMERICOS = ("int", "float")
ERRO = "erro"
NIVEIS = [("OR",), ("AND",), ("EQ", "NEQ"), ("LT", "LE", "GT", "GE"), ("PLUS", "MINUS"), ("STAR", "SLASH", "PERCENT")]


class ErroSintaxe(Exception):
    pass


class Scope:
    def __init__(self, nome: str, pai: "Scope | None"):
        self.nome = nome
        self.pai = pai
        self.simbolos: dict[str, tuple[str, int, int]] = {}

    def lookup(self, nome: str):
        s = self
        while s is not None:
            if nome in s.simbolos:
                return s.simbolos[nome]
            s = s.pai
        return None


@dataclass
class Resultado:
    programa: A.Programa | None
    diagnosticos: list[str]
    codigo: int


class Parser:
    def __init__(self, ts: TokenStream):
        self.ts = ts
        self.diags: list[str] = []
        self.blocos = 0

    # -- utilitários ------------------------------------------------------

    def _erro_sintaxe(self, token, esperado: str):
        encontrado = "fim da entrada" if token.type == "EOF" else f"'{token.lexeme}'"
        raise ErroSintaxe(
            f"Erro de sintaxe na linha {token.line}, coluna {token.column}: "
            f"esperado {esperado}; encontrado {encontrado}."
        )

    def _exige(self, tipo: str, esperado: str):
        if not self.ts.check(tipo):
            self._erro_sintaxe(self.ts.lookahead(), esperado)
        return self.ts.next_token()

    def _semantico(self, linha: int, coluna: int, mensagem: str):
        self.diags.append(f"Erro semântico na linha {linha}, coluna {coluna}: {mensagem}.")

    # -- programa, declarações e comandos --------------------------------

    def programa(self) -> A.Programa:
        escopo = Scope("global", None)
        itens = []
        while not self.ts.check("EOF"):
            itens += self.item(escopo)
        return A.Programa(tuple(itens))

    def item(self, escopo: Scope) -> list:
        if self.ts.lookahead().type in TIPOS:
            return self.decl(escopo)
        return [self.cmd(escopo)]

    def decl(self, escopo: Scope) -> list:
        tipo = TIPOS[self.ts.next_token().type]  # tipo.type (sintetizado)
        itens = [self.decl_item(tipo, escopo)]  # decl_item.inh = tipo.type
        while self.ts.accept("COMMA"):
            itens.append(self.decl_item(tipo, escopo))
        self._exige("SEMICOLON", "',' ou ';'")
        return itens

    def decl_item(self, inh: str, escopo: Scope) -> A.VarDecl:
        tok = self._exige("IDENT", "identificador")
        init = self.expr(escopo) if self.ts.accept("ASSIGN") else None
        if init is not None and not compativel(inh, init.tipo):
            self._semantico(
                init.linha, init.coluna, f"não é possível inicializar '{tok.lexeme}' ({inh}) com {init.tipo}"
            )
        if tok.lexeme in escopo.simbolos:
            _, l, c = escopo.simbolos[tok.lexeme]
            self._semantico(
                tok.line, tok.column,
                f"redeclaração de '{tok.lexeme}' no escopo {escopo.nome} (declarado na linha {l}, coluna {c})",
            )
        else:
            escopo.simbolos[tok.lexeme] = (inh, tok.line, tok.column)
        return A.VarDecl(tok.lexeme, inh, escopo.nome, init, tok.line, tok.column)

    def cmd(self, escopo: Scope):
        tok = self.ts.lookahead()
        if tok.type == "LBRACE":
            self.ts.next_token()
            self.blocos += 1
            interno = Scope(f"bloco{self.blocos}", escopo)
            itens = []
            while not self.ts.check("RBRACE"):
                if self.ts.check("EOF"):
                    self._erro_sintaxe(self.ts.lookahead(), "'}'")
                itens += self.item(interno)
            self.ts.next_token()
            return A.Block(interno.nome, tuple(itens), tok.line, tok.column)
        if tok.type in ("KW_IF", "KW_WHILE"):
            self.ts.next_token()
            self._exige("LPAREN", "'('")
            cond = self.expr(escopo)
            self._exige("RPAREN", "operador ou ')'")
            nome = "if" if tok.type == "KW_IF" else "while"
            if cond.tipo not in ("bool", ERRO):
                self._semantico(cond.linha, cond.coluna, f"condição do {nome} deve ser bool; recebeu {cond.tipo}")
            corpo = self.cmd(escopo)
            if tok.type == "KW_WHILE":
                return A.While(cond, corpo, tok.line, tok.column)
            senao = self.cmd(escopo) if self.ts.accept("KW_ELSE") else None
            return A.If(cond, corpo, senao, tok.line, tok.column)
        if tok.type == "KW_RETURN":
            self.ts.next_token()
            valor = None if self.ts.check("SEMICOLON") else self.expr(escopo)
            self._exige("SEMICOLON", "operador ou ';'")
            return A.Return(valor, tok.line, tok.column)
        if tok.type == "IDENT":
            alvo = self.identificador(self.ts.next_token(), escopo)
            self._exige("ASSIGN", "'='")
            valor = self.expr(escopo)
            self._exige("SEMICOLON", "operador ou ';'")
            if ERRO not in (alvo.tipo, valor.tipo) and not compativel(alvo.tipo, valor.tipo):
                self._semantico(
                    valor.linha, valor.coluna, f"não é possível atribuir {valor.tipo} a '{alvo.nome}' ({alvo.tipo})"
                )
            return A.Assign(alvo, valor, alvo.linha, alvo.coluna)
        self._erro_sintaxe(tok, "comando")

    # -- expressões (ações calculam o atributo sintetizado `tipo`) ----------

    def expr(self, escopo: Scope, nivel: int = 0) -> A.Expr:
        if nivel == len(NIVEIS):
            return self.unaria(escopo)
        no = self.expr(escopo, nivel + 1)
        while self.ts.check(*NIVEIS[nivel]):
            op = self.ts.next_token()
            direita = self.expr(escopo, nivel + 1)
            tipo = self.tipo_binario(op, no.tipo, direita.tipo)
            no = A.Binary(op.lexeme, no, direita, tipo, op.line, op.column)
        return no

    def unaria(self, escopo: Scope) -> A.Expr:
        op = self.ts.accept("MINUS", "NOT")
        if op is None:
            return self.primaria(escopo)
        o = self.unaria(escopo)
        tipo = ERRO
        if o.tipo == ERRO:
            pass
        elif op.lexeme == "-" and o.tipo in NUMERICOS:
            tipo = o.tipo
        elif op.lexeme == "!" and o.tipo == "bool":
            tipo = "bool"
        else:
            exigido = "numérico" if op.lexeme == "-" else "bool"
            self._semantico(op.line, op.column, f"operador '{op.lexeme}' exige operando {exigido}; recebeu {o.tipo}")
        return A.Unary(op.lexeme, o, tipo, op.line, op.column)

    def primaria(self, escopo: Scope) -> A.Expr:
        tok = self.ts.lookahead()
        literais = {"INT": "int", "FLOAT": "float", "KW_TRUE": "bool", "KW_FALSE": "bool"}
        if tok.type in literais:
            self.ts.next_token()
            return A.Literal(tok.lexeme, literais[tok.type], tok.line, tok.column)
        if tok.type == "IDENT":
            return self.identificador(self.ts.next_token(), escopo)
        if self.ts.accept("LPAREN"):
            e = self.expr(escopo)
            self._exige("RPAREN", "operador ou ')'")
            return e
        self._erro_sintaxe(tok, "expressão")

    def identificador(self, tok, escopo: Scope) -> A.Id:
        decl = escopo.lookup(tok.lexeme)
        if decl is None:
            self._semantico(tok.line, tok.column, f"identificador '{tok.lexeme}' não declarado")
            return A.Id(tok.lexeme, ERRO, tok.line, tok.column)
        return A.Id(tok.lexeme, decl[0], tok.line, tok.column)

    def tipo_binario(self, op, a: str, b: str) -> str:
        if ERRO in (a, b):
            return ERRO
        s = op.lexeme
        nums = a in NUMERICOS and b in NUMERICOS
        if s in ("+", "-", "*", "/"):
            if nums:
                return "float" if "float" in (a, b) else "int"
            exigido = "numéricos"
        elif s == "%":
            if a == b == "int":
                return "int"
            exigido = "int"
        elif s in ("<", "<=", ">", ">="):
            if nums:
                return "bool"
            exigido = "numéricos"
        elif s in ("==", "!="):
            if nums or a == b == "bool":
                return "bool"
            exigido = "do mesmo tipo (numéricos ou bool)"
        else:
            if a == b == "bool":
                return "bool"
            exigido = "bool"
        self._semantico(op.line, op.column, f"operador '{s}' exige operandos {exigido}; recebeu {a} e {b}")
        return ERRO


def compativel(destino: str, origem: str) -> bool:
    return origem == ERRO or destino == origem or (destino == "float" and origem == "int")


def parse(source: str) -> Resultado:
    """Pipeline completo: scanner -> parser/ações -> AST com atributos."""
    ts, erros_lexicos = scan(source)
    if erros_lexicos:
        return Resultado(None, [str(e) for e in erros_lexicos], ERRO_LEXICO)
    p = Parser(ts)
    try:
        programa = p.programa()
    except ErroSintaxe as erro:
        return Resultado(None, [str(erro)], ERRO_SINTAXE)
    return Resultado(programa, p.diags, ERRO_SEMANTICO if p.diags else OK)


def main(argv: list[str]) -> int:
    sys.stdout.reconfigure(encoding="utf-8", errors="surrogateescape")
    sys.stderr.reconfigure(encoding="utf-8", errors="surrogateescape")
    if len(argv) > 2:
        print("uso: python parser.py [arquivo | -]", file=sys.stderr)
        return ERRO_USO
    caminho = argv[1] if len(argv) == 2 else "-"
    try:
        dados = sys.stdin.buffer.read() if caminho == "-" else Path(caminho).read_bytes()
    except OSError as exc:
        print(f"erro ao abrir '{caminho}': {exc.strerror}", file=sys.stderr)
        return ERRO_USO

    r = parse(dados.decode("utf-8", errors="surrogateescape"))
    if r.programa is not None:
        print("\n".join(A.serializa(r.programa)))
    for d in r.diagnosticos:
        print(d, file=sys.stderr)
    return r.codigo


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: exercicios/aula13/ex10_pipeline/parser.c
/* Implementação de parser.h — mesma gramática e mesmas ações de parser.py.
 *
 * O erro de sintaxe interrompe a análise inteira com longjmp (o equivalente
 * da exceção ErroSintaxe do Python). Nada vaza nesse salto: os nós estão na
 * arena do Programa e os escopos numa lista do Parser, e os dois são
 * liberados por quem instalou o setjmp.
 */

#include "parser.h"

#include <setjmp.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    char *nome;
    Tipo tipo;
    int linha, coluna;
} Simbolo;

typedef struct Scope {
    char *nome;
    struct Scope *pai;
    Simbolo *v;
    size_t n, cap;
} Scope;

typedef struct {
    TokenStream *ts;
    Programa *prog;
    Diagnosticos *diags;
    Scope **escopos; /* todos os escopos criados (liberados no fim) */
    size_t n_escopos, cap_escopos;
    int blocos;
    jmp_buf sintaxe;
} Parser;

/* -- utilitários ------------------------------------------------------------ */

static void diag(Diagnosticos *d, char *msg) {
    if (d->n == d->cap) {
        d->cap = d->cap ? d->cap * 2 : 8;
        d->v = xrealloc(d->v, d->cap * sizeof(char *));
    }
    d->v[d->n++] = msg;
}

void diagnosticos_free(Diagnosticos *d) {
    for (size_t i = 0; i < d->n; i++) free(d->v[i]);
    free(d->v);
    *d = (Diagnosticos){0};
}

static void semantico(Parser *p, int linha, int coluna, char *msg) {
    diag(p->diags, xformat("Erro semântico na linha %d, coluna %d: %s.", linha, coluna, msg));
    free(msg);
}

static void erro_sintaxe(Parser *p, const Token *t, const char *esperado) {
    char *encontrado = t->type == TOK_EOF ? xstrdup("fim da entrada") : xformat("'%s'", t->lexeme);
    diag(p->diags, xformat("Erro de sintaxe na linha %d, coluna %d: esperado %s; encontrado %s.", t->line, t->column,
                           esperado, encontrado));
    free(encontrado);
    longjmp(p->sintaxe, 1);
}

static const Token *exige(Parser *p, TokenType tipo, const char *esperado) {
    if (!ts_check(p->ts, tipo)) erro_sintaxe(p, ts_lookahead(p->ts, 0), esperado);
    return ts_next_token(p->ts);
}

static Scope *novo_escopo(Parser *p, char *nome, Scope *pai) {
    Scope *s = xmalloc(sizeof(Scope));
    *s = (Scope){nome, pai, NULL, 0, 0};
    if (p->n_escopos == p->cap_escopos) {
        p->cap_escopos = p->cap_escopos ? p->cap_escopos * 2 : 8;
        p->escopos = xrealloc(p->escopos, p->cap_escopos * sizeof(Scope *));
    }
    p->escopos[p->n_escopos++] = s;
    return s;
}

static Simbolo *procura_local(Scope *s, const char *nome) {
    for (size_t i = 0; i < s->n; i++) {
        if (strcmp(s->v[i].nome, nome) == 0) return &s->v[i];
    }
    return NULL;
}

static Simbolo *lookup(Scope *s, const char *nome) {
    for (; s; s = s->pai) {
        Simbolo *sim = procura_local(s, nome);
        if (sim) return sim;
    }
    return NULL;
}

static int compativel(Tipo destino, Tipo origem) {
    return origem == TY_ERRO || destino == origem || (destino == TY_FLOAT && origem == TY_INT);
}

static int numerico(Tipo t) { return t == TY_INT || t == TY_FLOAT; }

static Tipo tipo_de_token(TokenType t) {
    return t == TOK_KW_INT ? TY_INT : t == TOK_KW_FLOAT ? TY_FLOAT : t == TOK_KW_BOOL ? TY_BOOL : TY_ERRO;
}

/* -- expressões ----------------------------------------------------------------- */

static const TokenType NIVEIS[][4] = {
    {TOK_OR}, {TOK_AND}, {TOK_EQ, TOK_NEQ}, {TOK_LT, TOK_LE, TOK_GT, TOK_GE}, {TOK_PLUS, TOK_MINUS},
    {TOK_STAR, TOK_SLASH, TOK_PERCENT},
};
static const int N_POR_NIVEL[] = {1, 1, 2, 4, 2, 3};
#define N_NIVEIS 6

static Expr *expr(Parser *p, Scope *escopo, int nivel);

static Expr *identificador(Parser *p, const Token *t, Scope *escopo) {
    Simbolo *s = lookup(escopo, t->lexeme);
    if (!s) semantico(p, t->line, t->column, xformat("identificador '%s' não declarado", t->lexeme));
    return expr_novo(&p->prog->arena, EX_ID, t->lexeme, s ? s->tipo : TY_ERRO, t->line, t->column, NULL, NULL);
}

static Tipo tipo_binario(Parser *p, const Token *op, Tipo a, Tipo b) {
    if (a == TY_ERRO || b == TY_ERRO) return TY_ERRO;
    const char *s = op->lexeme;
    const char *exigido;
    int nums = numerico(a) && numerico(b);
    if (strcmp(s, "+") == 0 || strcmp(s, "-") == 0 || strcmp(s, "*") == 0 || strcmp(s, "/") == 0) {
        if (nums) return (a == TY_FLOAT || b == TY_FLOAT) ? TY_FLOAT : TY_INT;
        exigido = "numéricos";
    } else if (strcmp(s, "%") == 0) {
        if (a == TY_INT && b == TY_INT) return TY_INT;
        exigido = "int";
    } else if (strcmp(s, "<") == 0 || strcmp(s, "<=") == 0 || strcmp(s, ">") == 0 || strcmp(s, ">=") == 0) {
        if (nums) return TY_BOOL;
        exigido = "numéricos";
    } else if (strcmp(s, "==") == 0 || strcmp(s, "!=") == 0) {
        if (nums || (a == TY_BOOL && b == TY_BOOL)) return TY_BOOL;
        exigido = "do mesmo tipo (numéricos ou bool)";
    } else {
        if (a == TY_BOOL && b == TY_BOOL) return TY_BOOL;
        exigido = "bool";
    }
    semantico(p, op->line, op->column,
              xformat("operador '%s' exige operandos %s; recebeu %s e %s", s, exigido, NOME_TIPO[a], NOME_TIPO[b]));
    return TY_ERRO;
}

static Expr *primaria(Parser *p, Scope *escopo) {
    const Token *t = ts_lookahead(p->ts, 0);
    Tipo lit = t->type == TOK_INT                                ? TY_INT
               : t->type == TOK_FLOAT                            ? TY_FLOAT
               : (t->type == TOK_KW_TRUE || t->type == TOK_KW_FALSE) ? TY_BOOL
                                                                  : TY_ERRO;
    if (lit != TY_ERRO) {
        ts_next_token(p->ts);
        return expr_novo(&p->prog->arena, EX_LITERAL, t->lexeme, lit, t->line, t->column, NULL, NULL);
    }
    if (t->type == TOK_IDENT) return identificador(p, ts_next_token(p->ts), escopo);
    if (ts_accept(p->ts, TOK_LPAREN)) {
        Expr *e = expr(p, escopo, 0);
        exige(p, TOK_RPAREN, "operador ou ')'");
        return e;
    }
    erro_sintaxe(p, t, "expressão");
    return NULL; /* inalcançável */
}

static Expr *unaria(Parser *p, Scope *escopo) {
    const Token *op = ts_accept(p->ts, TOK_MINUS);
    if (!op) op = ts_accept(p->ts, TOK_NOT);
    if (!op) return primaria(p, escopo);
    Expr *o = unaria(p, escopo);
    Tipo tipo = TY_ERRO;
    if (o->tipo == TY_ERRO) {
        /* erro já reportado */
    } else if (op->type == TOK_MINUS && numerico(o->tipo)) {
        tipo = o->tipo;
    } else if (op->type == TOK_NOT && o->tipo == TY_BOOL) {
        tipo = TY_BOOL;
    } else {
        semantico(p, op->line, op->column,
                  xformat("operador '%s' exige operando %s; recebeu %s", op->lexeme,
                          op->type == TOK_MINUS ? "numérico" : "bool", NOME_TIPO[o->tipo]));
    }
    return expr_novo(&p->prog->arena, EX_UNARY, op->lexeme, tipo, op->line, op->column, o, NULL);
}

static int no_nivel(TokenType t, int nivel) {
    for (int i = 0; i < N_POR_NIVEL[nivel]; i++) {
        if (NIVEIS[nivel][i] == t) return 1;
    }
    return 0;
}

static Expr *expr(Parser *p, Scope *escopo, int nivel) {
    if (nivel == N_NIVEIS) return unaria(p, escopo);
    Expr *e = expr(p, escopo, nivel + 1);
    while (no_nivel(ts_lookahead(p->ts, 0)->type, nivel)) {
        const Token *op = ts_next_token(p->ts);
        Expr *dir = expr(p, escopo, nivel + 1);
        Tipo tipo = tipo_binario(p, op, e->tipo, dir->tipo);
        e = expr_novo(&p->prog->arena, EX_BINARY, op->lexeme, tipo, op->line, op->column, e, dir);
    }
    return e;
}

/* -- declarações e comandos ------------------------------------------------------ */

static void item(Parser *p, Scope *escopo, No *bloco);

/* decl_item -> id [ = expr ]; `inh` é o tipo herdado da declaração. */
static No *decl_item(Parser *p, Tipo inh, Scope *escopo) {
    const Token *t = exige(p, TOK_IDENT, "identificador");
    Expr *init = ts_accept(p->ts, TOK_ASSIGN) ? expr(p, escopo, 0) : NULL;
    if (init && !compativel(inh, init->tipo)) {
        semantico(p, init->linha, init->coluna,
                  xformat("não é possível inicializar '%s' (%s) com %s", t->lexeme, NOME_TIPO[inh],
                          NOME_TIPO[init->tipo]));
    }
    Simbolo *ant = procura_local(escopo, t->lexeme);
    if (ant) {
        semantico(p, t->line, t->column,
                  xformat("redeclaração de '%s' no escopo %s (declarado na linha %d, coluna %d)", t->lexeme,
                          escopo->nome, ant->linha, ant->coluna));
    } else {
        if (escopo->n == escopo->cap) {
            escopo->cap = escopo->cap ? escopo->cap * 2 : 8;
            escopo->v = xrealloc(escopo->v, escopo->cap * sizeof(Simbolo));
        }
        escopo->v[escopo->n++] = (Simbolo){xstrdup(t->lexeme), inh, t->line, t->column};
    }
    No *d = no_novo(&p->prog->arena, ND_VARDECL, t->line, t->column);
    d->nome = xstrdup(t->lexeme);
    d->tipo = inh;
    d->escopo = xstrdup(escopo->nome);
    d->e1 = init;
    return d;
}

static No *cmd(Parser *p, Scope *escopo) {
    const Token *t = ts_lookahead(p->ts, 0);
    Arena *ar = &p->prog->arena;
    switch (t->type) {
    case TOK_LBRACE: {
        ts_next_token(p->ts);
        Scope *interno = novo_escopo(p, xformat("bloco%d", ++p->blocos), escopo);
        No *b = no_novo(ar, ND_BLOCK, t->line, t->column);
        b->escopo = xstrdup(interno->nome);
        while (!ts_check(p->ts, TOK_RBRACE)) {
            if (ts_check(p->ts, TOK_EOF)) erro_sintaxe(p, ts_lookahead(p->ts, 0), "'}'");
            item(p, interno, b);
        }
        ts_next_token(p->ts);
        return b;
    }
    case TOK_KW_IF:
    case TOK_KW_WHILE: {
        ts_next_token(p->ts);
        exige(p, TOK_LPAREN, "'('");
        Expr *cond = expr(p, escopo, 0);
        exige(p, TOK_RPAREN, "operador ou ')'");
        const char *nome = t->type == TOK_KW_IF ? "if" : "while";
        if (cond->tipo != TY_BOOL && cond->tipo != TY_ERRO) {
            semantico(p, cond->linha, cond->coluna,
                      xformat("condição do %s deve ser bool; recebeu %s", nome, NOME_TIPO[cond->tipo]));
        }
        No *n = no_novo(ar, t->type == TOK_KW_IF ? ND_IF : ND_WHILE, t->line, t->column);
        n->e1 = cond;
        n->c1 = cmd(p, escopo);
        if (t->type == TOK_KW_IF && ts_accept(p->ts, TOK_KW_ELSE)) n->c2 = cmd(p, escopo);
        return n;
    }
    case TOK_KW_RETURN: {
        ts_next_token(p->ts);
        No *n = no_novo(ar, ND_RETURN, t->line, t->column);
        if (!ts_check(p->ts, TOK_SEMICOLON)) n->e1 = expr(p, escopo, 0);
        exige(p, TOK_SEMICOLON, "operador ou ';'");
        return n;
    }
    case TOK_IDENT: {
        Expr *alvo = identificador(p, ts_next_token(p->ts), escopo);
        exige(p, TOK_ASSIGN, "'='");
        Expr *valor = expr(p, escopo, 0);
        exige(p, TOK_SEMICOLON, "operador ou ';'");
        if (alvo->tipo != TY_ERRO && valor->tipo != TY_ERRO && !compativel(alvo->tipo, valor->tipo)) {
            semantico(p, valor->linha, valor->coluna,
                      xformat("não é possível atribuir %s a '%s' (%s)", NOME_TIPO[valor->tipo], alvo->texto,
                              NOME_TIPO[alvo->tipo]));
        }
        No *n = no_novo(ar, ND_ASSIGN, alvo->linha, alvo->coluna);
        n->e1 = alvo;
        n->e2 = valor;
        return n;
    }
    default: erro_sintaxe(p, t, "comando"); return NULL; /* inalcançável */
    }
}

/* item -> decl | cmd; acrescenta os nós em `bloco` (ou no programa, se NULL). */
static void item(Parser *p, Scope *escopo, No *bloco) {
    Tipo tipo = tipo_de_token(ts_lookahead(p->ts, 0)->type);
    if (tipo == TY_ERRO) {
        No *c = cmd(p, escopo);
        if (bloco) no_add(bloco, c);
        else programa_add(p->prog, c);
        return;
    }
    ts_next_token(p->ts); /* tipo.type, herdado por cada decl_item */
    do {
        No *d = decl_item(p, tipo, escopo);
        if (bloco) no_add(bloco, d);
        else programa_add(p->prog, d);
    } while (ts_accept(p->ts, TOK_COMMA));
    exige(p, TOK_SEMICOLON, "',' ou ';'");
}

int parse(TokenStream *ts, Programa **out, Diagnosticos *diags) {
    Parser *p = xmalloc(sizeof(Parser)); /* no heap: sobrevive intacto ao longjmp */
    *p = (Parser){.ts = ts, .diags = diags};
    p->prog = xmalloc(sizeof(Programa));
    *p->prog = (Programa){0};
    int codigo;
    if (setjmp(p->sintaxe) == 0) {
        Scope *global = novo_escopo(p, xstrdup("global"), NULL);
        while (!ts_check(ts, TOK_EOF)) item(p, global, NULL);
        *out = p->prog;
        codigo = diags->n ? ERRO_SEMANTICO : OK;
    } else {
        /* Erro de sintaxe: descarta o programa parcial e os erros semânticos
         * vistos até aqui; fica só a mensagem de sintaxe (a última). */
        programa_free(p->prog);
        char *msg = diags->v[diags->n - 1];
        for (size_t i = 0; i + 1 < diags->n; i++) free(diags->v[i]);
        diags->v[0] = msg;
        diags->n = 1;
        *out = NULL;
        codigo = ERRO_SINTAXE;
    }
    for (size_t i = 0; i < p->n_escopos; i++) {
        Scope *s = p->escopos[i];
        for (size_t k = 0; k < s->n; k++) free(s->v[k].nome);
        free(s->v), free(s->nome), free(s);
    }
    free(p->escopos);
    free(p);
    return codigo;
}
